contents/revisionControl/bestPracticesGit.md

```diff
@@ -0,0 +1,259 @@
+
+### What Constitutes a Good Commit Message
+The easiest way to attain commit message discipline is to stop putting in one-liner descriptions using `git commit -m "Add some things to that."`. Instead, use `git commit` and write a proper message in an editor.
+
+A good commit message can be formatted the following way:
+```
+Capitalized, short (50 chars or less) summary
+
+More detailed explanatory text, if necessary.  Wrap it to about 72
+characters or so.  In some contexts, the first line is treated as the
+subject of an email and the rest of the text as the body.  The blank
+line separating the summary from the body is critical (unless you omit
+the body entirely); tools like rebase can get confused if you run the
+two together.
+
+Write your commit message in the imperative: "Fix bug" and not "Fixed bug"
+or "Fixes bug."  This convention matches up with commit messages generated
+by commands like git merge and git revert.
+
+Further paragraphs come after blank lines.
+
+- Bullet points are okay, too
+
+- Typically a hyphen or asterisk is used for the bullet, followed by a
+  single space, with blank lines in between, but conventions vary here
+
+- Use a hanging indent
+```
+
+Source: [A Note About Git Commit Messages by Tim Pope.](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)
+
+**Some Examples**
+- [torvalds/subsurfance-for-dirk](https://github.com/torvalds/subsurface-for-dirk/commits/master)
+- [torvalds/linux](https://github.com/torvalds/linux/commits/master)
+
+### Set Up Your Editor for Commit Messages
+1. To use your editor of choice for `git`-related functionality, e.g. `vim`, do one of either
+```
+git config --global core.editor "vim" # or you can do the following
+export GIT_EDITOR=vim # add to your .bashrc
+```
+2. Set your editor to wrap after 72 characters. In `vim`, you can do this by adding this to your `.vimrc`:
+```
+syntax on
+autocmd Filetype gitcommit spell textwidth=72
+```
+---
+## Always Commit Functional Code
+Merges to the following must always leave the project in a working state, i.e. it can be built and run on:
+- `master` branch, or equivalent,
+- `staging` branch, `development` branch or equivalent, if any.
+
+Changes to your own branches _that no one else is using_ can have non-functioning commits. However, you may wish to [hide the sausage making](#hide-the-sausage-making) to squash non-functioning commits into a single, functioning commit before you make a pull request.
+
+Changes to your own branches that is used by others should obey always-functioning-commits rule to minimize surprise. This is especially important if you expect your branch to be cherry-picked by another collaborator because they require a specific bit of code that you wrote.
+
+### Stashing
+If you need to switch between branches while in the middle of developing a commit, you can use the `git stash` command instead of committing your half-done code.
```

Comment:
Maybe what stash does should be explicitly stated?
you can use the `git stash` command **to save your code** instead of...